File: 92-Reverse-Linked-List-II/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        if(m==n) return head;
        ListNode *nodeOfM, *nodeOfN;
        nodeOfM = nodeOfN = head;
        ListNode *preNodeOfM = NULL;
        if(m>1) {
            m--;
            preNodeOfM = head;
            while(--m) preNodeOfM = preNodeOfM->next;
            nodeOfM = preNodeOfM->next;
        }
        while(--n) nodeOfN = nodeOfN->next;
        if(preNodeOfM) preNodeOfM->next = nodeOfN;
        else head = nodeOfN;
        ListNode *preNode, *curNode, *nextNode;
        preNode = nodeOfM;
        curNode = preNode->next;
        nodeOfM->next = nodeOfN->next;
        while(preNode!=nodeOfN) {
            nextNode = curNode->next;
            curNode->next = preNode;
            preNode = curNode;
            curNode = nextNode;
        }
        return head;
    }
};

File: 92-Reverse-Linked-List-II/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        if (head == null || head.next == null || m == n) {
            return head;
        }
        ListNode dummy = new ListNode(0);
        dummy.next = head;

        ListNode tail = dummy;
        while (m > 1) {
            tail = tail.next;
            m--;
            n--;
        }

        ListNode subTail = tail.next, subHead = subTail.next;
        ListNode prev = subTail, next = subHead.next;
        while (n > 1) {
            subHead.next = prev;
            prev = subHead;
            subHead = next;
            if (next != null) {
                next = next.next;
            }
            n--;
        }

        tail.next = prev;
        subTail.next = subHead;

        return dummy.next;
    }
}